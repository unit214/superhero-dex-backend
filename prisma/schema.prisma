// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Token {
  id         Int     @id @default(autoincrement())
  address    String  @unique
  symbol     String  @default("n/a")
  name       String  @default("n/a")
  decimals   Int     @default(-1)
  malformed  Boolean @default(false)
  noContract Boolean @default(false)
  pairs0     Pair[]  @relation("Pair_Token0")
  pairs1     Pair[]  @relation("Pair_Token1")
  listed     Boolean @default(false)
}

model Pair {
  id                          Int                               @id @default(autoincrement())
  address                     String                            @unique
  token0                      Token                             @relation("Pair_Token0", fields: [t0], references: [id])
  t0                          Int
  token1                      Token                             @relation("Pair_Token1", fields: [t1], references: [id])
  t1                          Int
  liquidityInfo               PairLiquidityInfo?                @relation("Pair_Info")
  synchronized                Boolean
  liquidityInfoHistory        PairLiquidityInfoHistory[]
  liquidityInfoHistoryV2      PairLiquidityInfoHistoryV2[]
  liquidityInfoHistoryError   PairLiquidityInfoHistoryError[]
  liquidityInfoHistoryV2Error PairLiquidityInfoHistoryV2Error[]
}

model PairLiquidityInfo {
  id          Int    @id
  pair        Pair   @relation("Pair_Info", fields: [id], references: [id])
  totalSupply String
  reserve0    String
  reserve1    String
  height      Int
}

model PairLiquidityInfoHistory {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  totalSupply    String
  reserve0       String
  reserve1       String
  height         Int
  microBlockHash String
  microBlockTime BigInt
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashUniqueIndex", [pairId, microBlockHash])
}

model PairLiquidityInfoHistoryError {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashErrorUniqueIndex", [pairId, microBlockHash, error])
}

model PairLiquidityInfoHistoryV2 {
  id              Int      @id @default(autoincrement())
  pair            Pair     @relation(fields: [pairId], references: [id])
  pairId          Int
  eventType       String
  height          Int
  microBlockTime  BigInt
  logIndex        Int
  microBlockHash  String
  transactionHash String
  totalSupply     String?
  reserve0        String?
  reserve1        String?
  deltaReserve0   String?
  deltaReserve1   String?
  fiatPrice       String?
  createdAt       DateTime @default(now())
  updatedAt       DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex])
}

model PairLiquidityInfoHistoryV2Error {
  id             Int      @id @default(autoincrement())
  pair           Pair     @relation(fields: [pairId], references: [id])
  pairId         Int
  microBlockHash String
  logIndex       Int
  error          String
  timesOccurred  Int      @default(1)
  createdAt      DateTime @default(now())
  updatedAt      DateTime @default(now()) @updatedAt

  @@unique(name: "pairIdMicroBlockHashLogIndexUniqueIndex", [pairId, microBlockHash, logIndex, error])
}
