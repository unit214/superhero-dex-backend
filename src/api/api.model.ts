import { ApiProperty } from '@nestjs/swagger';

// Patterns
export const contractPattern =
  'ct_([23456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz]){49,50}';
export const transactionPattern =
  'th_([23456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz]){49,50}';
export const accountPattern =
  'ak_([23456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz]){49,50}';
export const bigNumberPattern = '[1-9]+';
export const usdValuePattern = '[1-9]+([.][1-9]{1,4})?'; // [.] instead of \. needed, because of the \ being interpreted as an escape character in JavaScript resulting in incorrect pattern
export const aeValuePattern = '[1-9]+([.][1-9]{1,18})?'; // [.] instead of \. needed, because of the \ being interpreted as an escape character in JavaScript resulting in incorrect pattern
export const percentPattern = '[1-9]+([.][1-9]+)?'; // [.] instead of \. needed, because of the \ being interpreted as a escape character in JavaScript resulting in incorrect pattern
export const microBlockTimePattern = '[1-9]{13}';
export const microBlockHashPattern =
  'mh_([23456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz]){49,50}';

// Examples
export const heightExample = 598149;
export const bigNumberExample = '75819422625233931920';
export const usdValueExample = '0.1114';
export const aeValueExample = '6.251274413368500893';
export const percentExample = '26.579964573';

export const pairAddressPropertyOptions = {
  description: 'Pair contract address',
  pattern: contractPattern,
};
export const tokenAddressPropertyOptions = {
  description: 'Token contract address',
  pattern: contractPattern,
};

export class LiquidityInfo {
  @ApiProperty({
    description: 'Total supply of Liquidity Tokens',
    pattern: bigNumberPattern,
    example: bigNumberExample,
  })
  totalSupply: string;

  @ApiProperty({
    description: 'Whole reserve of token0 owned by the Pair contract',
    pattern: bigNumberPattern,
    example: bigNumberExample,
  })
  reserve0: string;

  @ApiProperty({
    description: 'Whole reserve of token1 owned by the Pair contract',
    pattern: bigNumberPattern,
    example: bigNumberExample,
  })
  reserve1: string;
}

export class PairBase {
  @ApiProperty(pairAddressPropertyOptions)
  address: string;

  @ApiProperty({
    description: `When service just started and it is in synchronization process or an error \
occurred regarding this pair, the pair is considered unsynced, therefore the user should not \
take in consideration it's liquidity info as the latest representation`,
    examples: [true, false],
  })
  synchronized: boolean;
}

export class Token {
  @ApiProperty({
    description: 'Token contract address',
    pattern: contractPattern,
  })
  address: string;

  @ApiProperty({
    description: 'Symbol which represents the token',
    examples: ['WAE', 'FST', 'SND', 'USDT'],
    example: 'WAE',
  })
  symbol: string;

  @ApiProperty({
    description: 'Full name of the token',
    examples: ['Wrapped AE', 'Token-A', 'Token-B', 'Wrapped Theter'],
    example: 'Wrapped AE',
  })
  name: string;

  @ApiProperty({
    description: `Total number of decimals represented by the token. \
If the token is marked as malformed or no contract was found decimals will be \`-1\``,
    example: 18,
  })
  decimals: number;

  @ApiProperty({
    description:
      "`true` in token meta-info (name/symbol/decimals/) could not be retrieved or doesn't follow AEX9 specifications",
    example: false,
  })
  malformed: boolean;

  @ApiProperty({
    description:
      '`true` if there is no contract deployed at the specified address',
    example: false,
  })
  noContract: boolean;
}

export class TokenWithListed extends Token {
  @ApiProperty({
    description:
      'Specifies if a token is supported/listed officially by the DEX or is just added by a random user',
    examples: [true, false],
  })
  listed: boolean;
}

const liquidityInfoPropertyOptions = {
  description: `Liquidity pair information. NOTE: between the pair addition moment and the first liquidity fetching \
liquidityInfo will be null. After that it will always have the last fetched values`,
  type: LiquidityInfo,
  nullable: true,
};

export class TokenWithPairAddresses extends TokenWithListed {
  @ApiProperty({
    description:
      'All pairs addresses in which a given token takes part (as token0 or as token1)',
    pattern: contractPattern,
  })
  pairs: string[];
}

export class TokenPairWithLiquidityInfo {
  @ApiProperty(pairAddressPropertyOptions)
  address: string;

  @ApiProperty({
    description: 'If pair is synchronized',
    examples: [true, false],
  })
  synchronized: boolean;

  @ApiProperty({ description: 'The other token from pair' })
  oppositeToken: TokenWithListed;

  @ApiProperty(liquidityInfoPropertyOptions)
  liquidityInfo?: LiquidityInfo;
}

export class TokenWithUsd extends TokenWithListed {
  @ApiProperty({
    description: 'Price of the token in AE',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  priceAe: string;

  @ApiProperty({
    description: 'Price of the token in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  priceUsd: string;

  @ApiProperty({
    description: 'Total locked value in AE',
    pattern: aeValuePattern,
    example: aeValueExample,
  })
  tvlAe: string;

  @ApiProperty({
    description: 'Total locked value in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  tvlUsd: string;

  @ApiProperty({
    description: 'Total Reserve',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  totalReserve: string;

  @ApiProperty({
    description: 'Number of pairs for this token',
    example: 16,
  })
  pairs: number;

  @ApiProperty({
    description: 'Volume for last day in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdDay: string;

  @ApiProperty({
    description: 'Volume for last week in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdWeek: string;

  @ApiProperty({
    description: 'Volume for last month in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdMonth: string;

  @ApiProperty({
    description: 'Volume for last year in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdYear: string;

  @ApiProperty({
    description: 'Volume for all time in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdAll: string;

  @ApiProperty({
    description: 'Price change for last day in percent',
    pattern: percentPattern,
    example: percentExample,
  })
  priceChangeDay: string;

  @ApiProperty({
    description: 'Price change for last week in percent',
    pattern: percentPattern,
    example: percentExample,
  })
  priceChangeWeek: string;

  @ApiProperty({
    description: 'Price change for last month in percent',
    pattern: percentPattern,
    example: percentExample,
  })
  priceChangeMonth: string;

  @ApiProperty({
    description: 'Price change for last year in percent',
    pattern: percentPattern,
    example: percentExample,
  })
  priceChangeYear: string;
}

export class PairWithUsd extends PairBase {
  @ApiProperty({
    description: 'Contract address for the token stored as token0',
    pattern: contractPattern,
  })
  token0: string;

  @ApiProperty({
    description: 'Contract address for the token stored as token1',
    pattern: contractPattern,
  })
  token1: string;

  @ApiProperty({
    description: 'How many transactions were made in the pair',
    example: 0,
  })
  transactions: number;

  @ApiProperty({
    description: 'Total Value Locked in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  tvlUsd: string;

  @ApiProperty({
    description: 'Volume for last day in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdDay: string;

  @ApiProperty({
    description: 'Volume for last week in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdWeek: string;

  @ApiProperty({
    description: 'Volume for last month in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdMonth: string;

  @ApiProperty({
    description: 'Volume for last year in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdYear: string;

  @ApiProperty({
    description: 'Volume for all time in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  volumeUsdAll: string;
}

class PairWithLiquidity extends PairBase {
  @ApiProperty(liquidityInfoPropertyOptions)
  liquidityInfo?: LiquidityInfo;
}
export class PairWithLiquidityAndTokens extends PairWithLiquidity {
  @ApiProperty({
    description: 'Token stored as token0',
    example: {
      address:
        process.env.DOC_TOKEN1 ||
        'ct_b7FZHQzBcAW4r43ECWpV3qQJMQJp5BxkZUGNKrqqLyjVRN3SC',
      symbol: 'WAE',
      name: 'Wrapped AE',
      decimals: 18,
    } as Token,
  })
  token0: Token;

  @ApiProperty({
    description: 'Token stored as token1',
    example: {
      address:
        process.env.DOC_TOKEN2 ||
        'ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF',
      symbol: 'USDT',
      name: 'Wrapped Theter',
      decimals: 18,
    } as Token,
  })
  token1: Token;
}

export class PairWithLiquidityAndTokenAddresses extends PairWithLiquidity {
  @ApiProperty({
    description: 'Contract address for the token stored as token0',
    pattern: contractPattern,
    example:
      process.env.DOC_TOKEN1 ||
      'ct_b7FZHQzBcAW4r43ECWpV3qQJMQJp5BxkZUGNKrqqLyjVRN3SC',
  })
  token0: string;

  @ApiProperty({
    description: 'Contract address for the token stored as token1',
    pattern: contractPattern,
    example:
      process.env.DOC_TOKEN2 ||
      'ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF',
  })
  token1: string;
}

export class TokenPairs {
  @ApiProperty({
    description: 'All the pairs which have as token0 the given token',
    type: [TokenPairWithLiquidityInfo],
  })
  pairs0: TokenPairWithLiquidityInfo[];

  @ApiProperty({
    description: 'All the pairs which have as token1 the given token',
    type: [TokenPairWithLiquidityInfo],
  })
  pairs1: TokenPairWithLiquidityInfo[];
}

export class GlobalState {
  @ApiProperty({
    description: 'Maximum block-height found in liquidity-info',
    example: heightExample,
  })
  topBlockHeight: number | null;

  @ApiProperty({
    description: 'How many pairs are synced (in percent)',
    example: 25.5,
  })
  pairsSyncedPercent: number;

  @ApiProperty({
    description: 'Total number of tokens',
    example: 7,
  })
  tokens: number;

  @ApiProperty({
    description: 'How many tokens are officially listed',
    example: 4,
  })
  listedTokens: number;

  @ApiProperty({
    description: 'Total number of pairs',
    example: 16,
  })
  pairs: number;

  @ApiProperty({
    description:
      'How many pairs contains only officially listed tokens (token0 & token1)',
    example: 7,
  })
  listedPairs: number;
}

export enum OrderQueryEnum {
  asc = 'asc',
  desc = 'desc',
}

class EventTypeEnum {
  static readonly CreatePair = 'CreatePair';
  static readonly PairMint = 'PairMint';
  static readonly PairBurn = 'PairBurn';
  static readonly SwapTokens = 'SwapTokens';
}

export class PairLiquidityInfoHistoryEntry {
  @ApiProperty({
    description: 'Pair Address of the history entry',
    pattern: contractPattern,
  })
  pairAddress: string;

  @ApiProperty({
    description: 'Type of the event',
    example: EventTypeEnum.SwapTokens,
  })
  type: EventTypeEnum;

  @ApiProperty({
    description: 'Whole reserve of token0 owned by the Pair contract',
    pattern: bigNumberPattern,
    example: bigNumberExample,
  })
  reserve0: string;

  @ApiProperty({
    description: 'Whole reserve of token1 owned by the Pair contract',
    pattern: bigNumberPattern,
    example: bigNumberExample,
  })
  reserve1: string;

  @ApiProperty({
    description: 'Change in reserve0',
    pattern: bigNumberPattern,
    example: bigNumberExample,
  })
  deltaReserve0: string;

  @ApiProperty({
    description: 'Change in reserve1',
    pattern: bigNumberPattern,
    example: bigNumberExample,
  })
  deltaReserve1: string;

  @ApiProperty({
    description: 'Exchange rate of token0 to AE',
    pattern: aeValuePattern,
    example: aeValueExample,
    type: 'string', // needed if type is string | null for the example to render correctly
  })
  token0AePrice: string | null;

  @ApiProperty({
    description: 'Exchange rate of token1 to AE',
    pattern: aeValuePattern,
    example: aeValueExample,
    type: 'string', // needed if type is string | null for the example to render correctly
  })
  token1AePrice: string | null;

  @ApiProperty({
    description: 'Price of AE in USD',
    pattern: usdValuePattern,
    example: usdValueExample,
  })
  aeUsdPrice: string;

  @ApiProperty({
    description: 'Block height of the history entry',
    example: heightExample,
  })
  height: number;

  @ApiProperty({
    description: 'Micro block hash of the history entry',
    pattern: microBlockHashPattern,
  })
  microBlockHash: string;

  @ApiProperty({
    description: 'Micro block time of the history entry',
    pattern: microBlockTimePattern,
  })
  microBlockTime: string;

  @ApiProperty({
    description: 'Transaction hash of the history entry',
    pattern: transactionPattern,
    example: 'th_JQzjTGpZGBtNtXMH1vYLwB1X2B6zUSLr6AfoQtgM6zQfXue4E',
  })
  transactionHash: string;

  @ApiProperty({
    description: 'Transaction index of the history entry',
    example: '0',
  })
  transactionIndex: string;

  @ApiProperty({
    description: 'Log index of the history entry',
    example: 0,
  })
  logIndex: number;

  @ApiProperty({
    description: 'Account that sent the transaction',
    pattern: accountPattern,
    example: 'ak_JQzjTGpZGBtNtXMH1vYLwB1X2B6zUSLr6AfoQtgM6zQfXue4E',
  })
  senderAccount: string;

  @ApiProperty({
    description: 'USD value of the reserve of token 0 of the pair',
    pattern: usdValuePattern,
    example: usdValueExample,
    type: 'string', // needed if type is string | null for the example to render correctly
  })
  reserve0Usd: string | null;

  @ApiProperty({
    description: 'USD value of the reserve of token 1 of the pair',
    pattern: usdValuePattern,
    example: usdValueExample,
    type: 'string', // needed if type is string | null for the example to render correctly
  })
  reserve1Usd: string | null;

  @ApiProperty({
    description: 'USD value of the transaction of token 0 of the pair',
    pattern: usdValuePattern,
    example: usdValueExample,
    type: 'string', // needed if type is string | null for the example to render correctly
  })
  delta0UsdValue: string | null;

  @ApiProperty({
    description: 'USD value of the transaction of token 1 of the pair',
    pattern: usdValuePattern,
    example: usdValueExample,
    type: 'string', // needed if type is string | null for the example to render correctly
  })
  delta1UsdValue: string | null;

  @ApiProperty({
    description: 'USD value of the pool fee',
    pattern: usdValuePattern,
    example: usdValueExample,
    type: 'string', // needed if type is string | null for the example to render correctly
  })
  txUsdFee: string | null;
}
