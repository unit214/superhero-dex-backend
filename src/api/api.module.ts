import { Module } from '@nestjs/common';

import { GraphController } from '@/api/graph/graph.controller';
import { GraphService } from '@/api/graph/graph.service';
import { PairLiquidityInfoHistoryController } from '@/api/pair-liquidity-info-history/pair-liquidity-info-history.controller';
import { PairLiquidityInfoHistoryService } from '@/api/pair-liquidity-info-history/pair-liquidity-info-history.service';
import { PairsController } from '@/api/pairs/pairs.controller';
import { PairsService } from '@/api/pairs/pairs.service';
import { SwapRoutesService } from '@/api/swap-routes/swap-route.service';
import { SwapRoutesController } from '@/api/swap-routes/swap-routes.controller';
import { TokensController } from '@/api/tokens/tokens.controller';
import { TokensService } from '@/api/tokens/tokens.service';
import { DatabaseModule } from '@/database/database.module';

@Module({
  imports: [DatabaseModule],
  controllers: [
    PairLiquidityInfoHistoryController,
    PairsController,
    TokensController,
    SwapRoutesController,
    GraphController,
  ],
  providers: [
    PairLiquidityInfoHistoryService,
    PairsService,
    TokensService,
    SwapRoutesService,
    GraphService,
  ],
})
export class ApiModule {}
