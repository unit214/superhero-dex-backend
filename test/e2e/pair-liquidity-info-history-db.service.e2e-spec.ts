import { PairLiquidityInfoHistoryDbService } from '../../src/database/pair-liquidity-info-history/pair-liquidity-info-history-db.service';
import { Test, TestingModule } from '@nestjs/testing';
import { PrismaService } from '../../src/database/prisma.service';
import { Pair, PairLiquidityInfoHistory, Token } from '@prisma/client';
import { OrderQueryEnum } from '../../src/api/api.model';

import { ContractAddress } from '../../src/clients/sdk-client.model';

const token1: Token = {
  id: 1,
  address: 'ct_token1',
  symbol: '1',
  name: '1',
  decimals: 18,
  malformed: false,
  noContract: false,
  listed: false,
};
const token2: Token = {
  id: 2,
  address: 'ct_token2',
  symbol: '2',
  name: '2',
  decimals: 18,
  malformed: false,
  noContract: false,
  listed: false,
};
const token3: Token = {
  id: 3,
  address: 'ct_token3',
  symbol: '3',
  name: '3',
  decimals: 18,
  malformed: false,
  noContract: false,
  listed: false,
};
const pair1: Pair = {
  id: 1,
  address: 'ct_pair1',
  t0: 1,
  t1: 2,
  synchronized: true,
};
const pair2: Pair = {
  id: 2,
  address: 'ct_pair2',
  t0: 2,
  t1: 3,
  synchronized: true,
};
const pair3: Pair = {
  id: 3,
  address: 'ct_pair4',
  t0: 2,
  t1: 3,
  synchronized: true,
};
const historyEntry1: PairLiquidityInfoHistory = {
  id: 1,
  pairId: 1,
  totalSupply: '2000148656239820912122563',
  reserve0: '950875688379385634428666',
  reserve1: '4208476309359648851631167',
  height: 100001,
  microBlockHash: 'mh_entry1',
  microBlockTime: 1000000000001n,
  updatedAt: new Date(),
};
const historyEntry2: PairLiquidityInfoHistory = {
  id: 2,
  pairId: 1,
  totalSupply: '9954575303087659158151',
  reserve0: '20210309618736130321327',
  reserve1: '4903471477408475598460',
  height: 200002,
  microBlockHash: 'mh_entry2',
  microBlockTime: 2000000000002n,
  updatedAt: new Date(),
};
const historyEntry3: PairLiquidityInfoHistory = {
  id: 3,
  pairId: 2,
  totalSupply: '56931443813890767374824',
  reserve0: '20556919390913460010617',
  reserve1: '157691178959228289022449',
  height: 300003,
  microBlockHash: 'mh_entry3',
  microBlockTime: 3000000000003n,
  updatedAt: new Date(),
};
const historyEntry4: PairLiquidityInfoHistory = {
  id: 4,
  pairId: 2,
  totalSupply: '56931443813890767374824',
  reserve0: '20556919390913460010617',
  reserve1: '157691178959228289022449',
  height: 300003,
  microBlockHash: 'mh_entry4',
  microBlockTime: 4000000000004n,
  updatedAt: new Date(),
};

describe('PairLiquidityInfoHistoryDbService', () => {
  let service: PairLiquidityInfoHistoryDbService;
  let prismaService: PrismaService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PairLiquidityInfoHistoryDbService, PrismaService],
    }).compile();

    service = module.get<PairLiquidityInfoHistoryDbService>(
      PairLiquidityInfoHistoryDbService,
    );
    prismaService = module.get<PrismaService>(PrismaService);
  });

  beforeEach(async () => {
    await prismaService.token.createMany({ data: [token1, token2, token3] });
    await prismaService.pair.createMany({ data: [pair1, pair2, pair3] });
    await prismaService.pairLiquidityInfoHistory.createMany({
      data: [historyEntry1, historyEntry2, historyEntry3, historyEntry4],
    });
  });

  afterEach(async () => {
    await prismaService.pairLiquidityInfoHistory.deleteMany();
    await prismaService.pair.deleteMany();
    await prismaService.token.deleteMany();
  });

  afterAll(async () => {
    await prismaService.$disconnect();
  });

  describe('getAll', () => {
    it('should return all entries', async () => {
      const result = await service.getAll(100, 0, OrderQueryEnum.asc);
      expect(result.map((e) => e.id)).toEqual([1, 2, 3, 4]);
    });

    it('should return return entries with limit, offset and order', async () => {
      const result = await service.getAll(2, 1, OrderQueryEnum.desc);
      expect(result.map((e) => e.id)).toEqual([3, 2]);
    });

    it('should correctly filter by pair address', async () => {
      const result = await service.getAll(
        100,
        0,
        OrderQueryEnum.asc,
        pair1.address as ContractAddress,
        undefined,
        undefined,
        undefined,
      );
      expect(result.map((e) => e.id)).toEqual([1, 2]);
    });

    it('should correctly filter by height', async () => {
      const result = await service.getAll(
        100,
        0,
        OrderQueryEnum.asc,
        undefined,
        300003,
        undefined,
        undefined,
      );
      expect(result.map((e) => e.id)).toEqual([3, 4]);
    });

    it('should correctly return entries newer or equal fromBlockTime', async () => {
      const result = await service.getAll(
        100,
        0,
        OrderQueryEnum.asc,
        undefined,
        undefined,
        2000000000002n,
        undefined,
      );
      expect(result.map((e) => e.id)).toEqual([2, 3, 4]);
    });

    it('should correctly return entries older or equal toBlockTime', async () => {
      const result = await service.getAll(
        100,
        0,
        OrderQueryEnum.desc,
        undefined,
        undefined,
        undefined,
        3000000000003n,
      );
      expect(result.map((e) => e.id)).toEqual([3, 2, 1]);
    });
  });

  describe('getLastlySyncedBlockByPairId', () => {
    it('should correctly return the last synced block for a given pairId', async () => {
      const result = await service.getLastlySyncedBlockByPairId(1);
      expect(result?.microBlockTime).toEqual(2000000000002n);
    });
  });

  describe('getWithinHeightSorted', () => {
    it('should correctly return all entries greater or equal a given height limit sorted by microBlockTime ascending', async () => {
      const result = await service.getWithinHeightSorted(200002);
      expect(result.map((e) => e.id)).toEqual([2, 3, 4]);
    });
  });

  describe('deleteFromMicroBlockTime', () => {
    it('should correctly delete all entries newer or equal a given block time', async () => {
      await service.deleteFromMicroBlockTime(3000000000003n);
      const result = await prismaService.pairLiquidityInfoHistory.findMany();
      expect(result.map((e) => e.id)).toEqual([1, 2]);
    });
  });
});
