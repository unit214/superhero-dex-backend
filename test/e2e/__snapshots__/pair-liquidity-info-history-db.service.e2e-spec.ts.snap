// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairLiquidityInfoHistoryDbService upsert should correctly upsert an existing entry 1`] = `
{
  "aeUsdPrice": "0.060559",
  "deltaReserve0": "-500",
  "deltaReserve1": "-500",
  "eventType": "PairBurn",
  "height": 200002,
  "id": 222,
  "logIndex": 1,
  "microBlockHash": "mh_entry2",
  "microBlockTime": 2000000000002n,
  "pairId": 1,
  "reserve0": "500",
  "reserve1": "500",
  "transactionHash": "th_entry2",
  "transactionIndex": 200002n,
}
`;
